### PR TITLE
Warm reboot: Orchagent state restore

### PR DESCRIPTION
**Pull request by jipanyang**
Signed-off-by: Jipan Yang <jipan.yang@alibaba-inc.com>

**What I did**
State restore for Orchagent warm start 

It includes basic after restore orchagent state validation: No un-met object dependency in orchagent.  

**Why I did it**

**How I verified it**
VS testing.  All existing vs test cases pass.

**Details if related**

Has dependency on:
https://github.com/Azure/sonic-swss-common/pull/211
https://github.com/Azure/sonic-swss-common/pull/214
https://github.com/Azure/sonic-swss/pull/546
https://github.com/Azure/sonic-swss/pull/547
https://github.com/Azure/sonic-swss/pull/551

Follow up PRs planned: 
<1> Orchagent  pre-warm-restart validation.
This is to have a deterministic known state before warm restart,  otherwise state restore will be a best effort processing.

<2> state sync up
dynamic data like port state, fdb, arp may change during the restart window.

<3> Warm restart phase state data.
This is to store basic warm restart state data as in  https://github.com/Azure/sonic-swss/pull/553
And more debug information should be provided like exact list of postponed tasks/executor during restore.

### CODE REVIEW

**Comment by lguohan** [comment]
need new vs test to test warm boot scenario

File: orchagent/aclorch.cpp [hunk 1]
```diff
@@ -1694,6 +1694,7 @@ void AclOrch::init(vector<TableConnector>& connectors, PortsOrch *portOrch, Mirr
     auto interv = timespec { .tv_sec = COUNTERS_READ_INTERVAL, .tv_nsec = 0 };
     auto timer = new SelectableTimer(interv);
     auto executor = new ExecutableTimer(timer, this);
+    executor->setName("ACL_POLL_TIMER");
```
Comment (lguohan) [inline orchagent/aclorch.cpp]: why not just change the constructor to set the name. 

not sure if it is useful to introduce setName method, since we do not (probably should not) allow change the name.

File: orchagent/orch.cpp [hunk 2]
```diff
@@ -171,6 +171,23 @@ void Consumer::drain()
...
+{
+    for (auto &tm :m_toSync)
+    {
+        KeyOpFieldsValuesTuple& tuple = tm.second;
+
+        string s = getTableName() + ":" + kfvKey(tuple)
+               + "|" + kfvOp(tuple);
+        for (auto i = kfvFieldsValues(tuple).begin(); i != kfvFieldsValues(tuple).end(); i++)
+        {
+            s += "|" + fvField(*i) + ":" + fvValue(*i);
```
Comment (lguohan) [inline orchagent/orch.cpp]: do not hardcode the separator.

File: orchagent/orchdaemon.cpp [hunk 3]
```diff
@@ -313,5 +358,81 @@ void OrchDaemon::start()
...
+{
+    /*
+     * No pending task should exist for any of the consumer at this point.
+     * All the prexisting data in appDB and configDb have been read and processed.
+     */
+    vector<string> ts;
+    getTaskToSync(ts);
+    if (ts.size() != 0)
+    {
+        // TODO: change it to fatal once staged ProducerStateTable/ConsumerStateTable change and
```
Comment (lguohan) [inline orchagent/orchdaemon.cpp]: remove TODO here?

File: orchagent/orchdaemon.cpp [hunk 4]
```diff
@@ -152,13 +162,13 @@ bool OrchDaemon::init()
         gAclOrch = new AclOrch(acl_table_connectors, gPortsOrch, mirror_orch, gNeighOrch, gRouteOrch);
     }
 
-    m_orchList.push_back(gAclOrch);
     m_orchList.push_back(gFdbOrch);
+    m_orchList.push_back(mirror_orch);
```
Comment (lguohan) [inline orchagent/orchdaemon.cpp]: why move mirror orch and acl orch after fdb orch?

File: orchagent/orchdaemon.cpp [hunk 5]
```diff
@@ -313,5 +358,81 @@ void OrchDaemon::start()
...
+                o->doTask();
+            }
+
+            warmRestoreValidation();
+            SWSS_LOG_NOTICE("Orchagent state restore done");
+            restored = true;
+            syncd_apply_view();
+
+            // TODO: should be set after port/fdb/arp sync up
+            WarmStart::setWarmStartState("orchagent", WarmStart::RECONCILED);
```
Comment (lguohan) [inline orchagent/orchdaemon.cpp]: how do we sync the port status and fdb notifications?

File: orchagent/orchdaemon.cpp (hunk 3 above)
Comment (jipanyang) [reply orchagent/orchdaemon.cpp]: For planned warm restart,  there is one PR https://github.com/Azure/sonic-swss/pull/562/files     for enforcing orchagent to be in certain state before stopping orchagent/swss docker.  After orchagent is restored, it should gain back the same state.

But there is open issue about the exact criteria of warm restart readiness. Currently we want to make sure that there is no un-met object dependency in orchagent which means the "SyncMap m_toSync;" of each Consumer is empty.   @zhenggen-xu  ever expressed concern over this check.

File: orchagent/orchdaemon.cpp (hunk 4 above)
Comment (jipanyang) [reply orchagent/orchdaemon.cpp]: This is to reduce the possible turbulence caused by the dependency relaxation processing during state restore.
mirror orch may use fdb mac/vlan info. 
   ``` if (!m_portsOrch->getPort(session.neighborInfo.neighbor.alias, session.neighborInfo.port))```
and acl orch may check mirror session
      ``` "if (!m_pMirrorOrch->getSessionState(m_sessionName, state))"```

File: orchagent/orchdaemon.cpp (hunk 5 above)
Comment (jipanyang) [reply orchagent/orchdaemon.cpp]: Port state sync: https://github.com/Azure/sonic-swss/pull/557
FDB state sync: https://github.com/Azure/sonic-swss/pull/558
I think @zhenggen-xu  is to propose PR for arp sync up.

File: orchagent/orchdaemon.cpp (hunk 3 above)
Comment (zhenggen-xu) [reply orchagent/orchdaemon.cpp]: This check is not necessary reach the desired result.  The problem is if some dependencies were never met (e,g, the nexthop is not resolved for some reason for a particular route), which was the case for pre-reboot,  but may kill the whole warm reboot process. Looks like we may want to handle those entries differently or we simply do a loose check (print warning etc) and move forward.

File: orchagent/orchdaemon.cpp (hunk 5 above)
Comment (lguohan) [reply orchagent/orchdaemon.cpp]: I feel we should take a different approach to solve port oper status and fdb sync in orchagent. 

Since we have information in asic db, we should just send syncd a command and ask syncd to enumerate all state in asic db and regenerate all the events back to orchagent. 

The ground truth is in asic_db.

Comment (jipanyang) [reply orchagent/orchdaemon.cpp]: We may have more in-depth discussion of port state sync up and fdb sync up in the corresponding PRs.  Here is some quick recap:

<1> Current SONiC use NotificationProducer to inform port state change and fdb,  which means orchagent will miss the signal if it happens between orchagent shutdown and startup.

Syncd doens't know whether the signal has been lost or not.  Also that notification event is triggered from libsai/SDK.

<2> In current port sync up implementation,  we do a get SAI_PORT_ATTR_OPER_STATUS via sai API to have the latest oper status, then update hostif and db accordingly.  

<3> For FDB sync up,   since syncd does put the latest FDB entry into ASIC DB,  yes, we are retrieving the latest FDB from ASIC.

Surely there is some other alternative, like change the Notification mechanism to some db data persistent channel. Under current framework, the above approaches have proven to be working and stable.  I should have put more comments to the port state and fdb sync up implementation.

syncd_apply_view() is called after state restore,  I guess that could make the life of syncd view comparison a little easier, but it is up to how you want view comparison to work.  The libsai redis idempotent API solution doesn't have dependency on  syncd_apply_view().

Comment (lguohan) [reply orchagent/orchdaemon.cpp]: we can ask syncd to pause sending notification when we are upgrading.

it is better to ask syncd to re-send the previous view to us. we do not orchagent to scan the asic view because the format might change and it can break orchagent code. orchagent should not look at the asic db.

Can you modify you fdb sync and port sync to syncd?

Comment (jipanyang) [reply orchagent/orchdaemon.cpp]: For port state sync, I don't think there is need to make the solution so complex.
FDB format could be a valid concern, but we should alway be careful when making format change.

My understanding of SONiC design about docker is to decouple the services and make them more independent from each other. But I'm seeing the intent to couple swss and syncd tightly together, what is the real purpose behind this?  Why were syncd and orchagent put into different dockers at the first place?

File: orchagent/orchdaemon.cpp (hunk 3 above)
Comment (jipanyang) [reply orchagent/orchdaemon.cpp]: Do we want to allow this scenario (there is unfulfilled dependency) for planned warm restart?

The possible alternative is to add a knob to have two levels of pre-reboot check:
1. Allow unfulfilled dependency.  
2. disallow unfulfilled dependency.  
After warm restore, we do the check accordingly.

With no intention to make it a hot topic here and block everything else,  I'll change the log level to NOTICE and continue for now.

Please go to https://github.com/Azure/sonic-swss/pull/562 for further discussion.

File: orchagent/orchdaemon.cpp [hunk 6]
```diff
@@ -313,5 +358,81 @@ void OrchDaemon::start()
...
+            /*
+             * drain remaining data that are out of order like LAG_MEMBER_TABLE and VLAN_MEMBER_TABLE
+             * since they were checked before LAG_TABLE and VLAN_TABLE.
+             */
+            for (Orch *o : m_orchList)
+            {
+                o->doTask();
+            }
+
+            warmRestoreValidation();
```
[the remark this answers is not in the document]
Comment (jipanyang) [reply orchagent/orchdaemon.cpp]: Yes,  once we reach agreement what should be checked before reboot, and enforce the validation after restore.

File: orchagent/orchdaemon.cpp [hunk 7]
```diff
@@ -304,7 +324,17 @@ void OrchDaemon::start()
         }
 
         auto *c = (Executor *)s;
-        c->execute();
+
+        if (restored)
```
Comment (qiluo-msft) [inline orchagent/orchdaemon.cpp]: >if [](start = 8, length = 2)

select() is used for normal processing, and it is not for warm start, because all the data is already added during bake() -> addExistingData().

You can extract the warm start logic and move before the normal processing while-loop.

Comment (jipanyang) [reply orchagent/orchdaemon.cpp]: Right.  Previously configDb data was retrieved here,  it is not needed any more.  pre-shutdown  warm restart check may need to be processed in select(), and possibly arp sync up event too,  but that could be done in separate PRs.

File: orchagent/orch.h [hunk 8]
```diff
@@ -68,9 +68,10 @@ class Orch;
 class Executor : public Selectable
 {
 public:
-    Executor(Selectable *selectable, Orch *orch)
+    Executor(Selectable *selectable, Orch *orch, const string &name)
         : m_selectable(selectable)
         , m_orch(orch)
+        , m_name(name)
```
Comment (qiluo-msft) [inline orchagent/orch.h]: >m_name [](start = 10, length = 6)

I believe the new name field is not necessary, and you could always get it from m_consumerMap key. Could you help explorer the possibility?

File: orchagent/orch.h [hunk 9]
```diff
@@ -116,15 +125,23 @@ class Consumer : public Executor {
...
     }
 
+    int getDbId() const
+    {
+        return getConsumerTable()->getDbId();
+    }
+
+    string dumpTuple(KeyOpFieldsValuesTuple &tuple);
+    void dumpToSyncTasks(vector<string> &ts);
+    void execute();
```
Comment (qiluo-msft) [inline orchagent/orch.h]: >void execute(); [](start = 4, length = 15)

No need to move this line.

File: orchagent/orch.h (hunk 8 above)
Comment (jipanyang) [reply orchagent/orch.h]: Isn't it a good practice to give executor a fixed unique name? For now it is mainly for debugging, later it may also be used for state validation log.  
I noticed that there are even cases where empty string given to to map as key in current code,  all those should be changed.

File: orchagent/orch.cpp [hunk 10]
```diff
@@ -185,6 +185,30 @@ void Consumer::drain()
         m_orch->doTask(*this);
 }
 
+string Consumer::dumpTuple(KeyOpFieldsValuesTuple &tuple)
+{
+    string s = getTableName() + getConsumerTable()->getTableNameSeparator() + kfvKey(tuple)
+               + "|" + kfvOp(tuple);
```
[the remark this answers is not in the document]
Comment (jipanyang) [reply orchagent/orch.cpp]: This a good question.
We had it fixed to ":" previously.  Changed it to getConsumerTable()->getTableNameSeparator() upon comment.  I think one of he benefits of doing this is we'll have the same key format as that in redis DB.  Any other idea?

File: orchagent/orch.cpp [hunk 11]
```diff
@@ -185,6 +185,30 @@ void Consumer::drain()
...
+    for (auto i = kfvFieldsValues(tuple).begin(); i != kfvFieldsValues(tuple).end(); i++)
+    {
+        s += "|" + fvField(*i) + ":" + fvValue(*i);
+    }
+
+    return s;
+}
+
+void Consumer::dumpToSyncTasks(vector<string> &ts)
```
Comment (qiluo-msft) [inline orchagent/orch.cpp]: >dumpToSyncTasks [](start = 15, length = 15)

dumpTuple() and dumpToSyncTasks() are not well designed interfaces.

I agree the observability is important.

File: orchagent/main.cpp [hunk 12]
```diff
@@ -22,6 +22,7 @@ extern "C" {
 #include "saihelper.h"
 #include "notifications.h"
 #include <signal.h>
+#include <warm_restart.h>
```
Comment (qiluo-msft) [inline orchagent/main.cpp]: #include "warm_restart.h"

File: orchagent/intfsorch.cpp [hunk 13]
```diff
@@ -86,7 +86,16 @@ void IntfsOrch::doTask(Consumer &consumer)
         {
             if (alias == "lo")
             {
-                addIp2MeRoute(ip_prefix);
+                // set request for lo may come after warm start restore
```
Comment (qiluo-msft) [inline orchagent/intfsorch.cpp]: Does it related to warm start? If not, let's do it in another PR.

File: orchagent/intfsorch.cpp [hunk 14]
```diff
@@ -171,6 +180,7 @@ void IntfsOrch::doTask(Consumer &consumer)
         {
             if (alias == "lo")
             {
+                m_syncdIntfses.erase(alias);
```
Comment (qiluo-msft) [inline orchagent/intfsorch.cpp]: The same.

File: orchagent/fdborch.cpp [hunk 15]
```diff
@@ -53,6 +53,14 @@ void FdbOrch::update(sai_fdb_event_t type, const sai_fdb_entry_t* entry, sai_obj
             return;
         }
 
+        // we already have such entries
+        if (m_entries.count(update.entry) != 0)
```
Comment (qiluo-msft) [inline orchagent/fdborch.cpp]: >count [](start = 22, length = 5)

find()

File: orchagent/crmorch.cpp [hunk 16]
```diff
@@ -162,8 +162,11 @@ CrmOrch::CrmOrch(DBConnector *db, string tableName):
         m_resourcesMap.emplace(res.first, CrmResourceEntry(res.second, CRM_THRESHOLD_TYPE_DEFAULT, CRM_THRESHOLD_LOW_DEFAULT, CRM_THRESHOLD_HIGH_DEFAULT));
     }
 
-    auto executor = new ExecutableTimer(m_timer.get(), this);
-    Orch::addExecutor("CRM_COUNTERS_POLL", executor);
+    // The CRM stats needs to be populated again
+    m_countersCrmTable->del(CRM_COUNTERS_TABLE_KEY);
```
Comment (qiluo-msft) [inline orchagent/crmorch.cpp]: >m_countersCrmTable->del(CRM_COUNTERS_TABLE_KEY); [](start = 4, length = 48)

If it is applied to cold start, let's do it in another PR.

File: orchagent/orch.h (hunk 8 above)
Comment (qiluo-msft) [reply orchagent/orch.h]: I agree on "even cases where empty string given to to map as key in current code, all those should be changed." and I will fix them.

File: orchagent/orch.h (hunk 9 above)
Comment (jipanyang) [reply orchagent/orch.h]: ok.

File: orchagent/orch.cpp [hunk 17]
```diff
@@ -185,6 +185,30 @@ void Consumer::drain()
...
+{
+    for (auto &tm :m_toSync)
```
[the remark this answers is not in the document]
Comment (jipanyang) [reply orchagent/orch.cpp]: ok

File: orchagent/crmorch.cpp (hunk 16 above)
Comment (jipanyang) [reply orchagent/crmorch.cpp]: For cold start,  no data exists for CRM stats table, so  it has no effect.

File: orchagent/fdborch.cpp (hunk 15 above)
Comment (jipanyang) [reply orchagent/fdborch.cpp]: ok.

File: orchagent/intfsorch.cpp (hunk 13 above)
Comment (jipanyang) [reply orchagent/intfsorch.cpp]: It is related to warm start.  It is noticed that sometime IP data for lo comes after warm start.

File: orchagent/main.cpp (hunk 12 above)
Comment (jipanyang) [reply orchagent/main.cpp]: ok.

File: orchagent/orch.cpp (hunk 11 above)
Comment (jipanyang) [reply orchagent/orch.cpp]: dumpTuple() is existing function name.  It is refactored so the common function may be used in multiple places. Any specific issue/suggestion?

File: orchagent/orch.cpp (hunk 17 above)
[the remark this answers is not in the document]
Comment (jipanyang) [reply orchagent/orch.cpp]: ok

File: orchagent/orch.h (hunk 9 above)
Comment (jipanyang) [reply orchagent/orch.h]: ok

File: orchagent/intfsorch.cpp (hunk 13 above)
Comment (qiluo-msft) [reply orchagent/intfsorch.cpp]: The logic is also applied to cold start. Are you fixing a bug also exists in cold start? if yes, we prefer a separate PR just for it.

---
In reply to: [209787586](https://github.com/Azure/sonic-swss/pull/554#discussion_r209787586) [](ancestors = 209787586)

File: orchagent/orch.h (hunk 8 above)
Comment (qiluo-msft) [reply orchagent/orch.h]: Fixed by [https://github.com/Azure/sonic-swss/pull/575](https://github.com/Azure/sonic-swss/pull/575)

---
In reply to: [209784061](https://github.com/Azure/sonic-swss/pull/554#discussion_r209784061) [](ancestors = 209784061)

Comment (jipanyang) [reply orchagent/orch.h]: I didn't say there was bug there, at least it is working. But it is a bad design putting empty string as key.
The purpose of name related change in this PR is to facilitate debugging and logging via a meaningful name for each excecutor including those timers and Notifiers.  I found it especially useful during the warm restart testing since very often you need to trace specific executor whenever problem happened.

File: orchagent/orch.cpp (hunk 11 above)
Comment (qiluo-msft) [reply orchagent/orch.cpp]: I will investigate and hopefully submit a PR

---
In reply to: [209787965](https://github.com/Azure/sonic-swss/pull/554#discussion_r209787965) [](ancestors = 209787965)

**Comment by jipanyang** [comment]
What exactly are you doing? Reviewing code? or copying code and making a little change then push your own pr and we have to merge it again and again?





来自钉钉专属商务邮箱------------------------------------------------------------------
Sender:Qi Luo<notifications@github.com>
Date:2018/08/15 02:07:03
Recipient:Azure/sonic-swss<sonic-swss@noreply.github.com>
Cc:YANG, Jipan(继攀)<jipan.yang@alibaba-inc.com>; Author<author@noreply.github.com>
Theme:Re: [Azure/sonic-swss] Warm reboot: Orchagent state restore (#554)

@qiluo-msft commented on this pull request. 
In orchagent/orch.cpp:
> @@ -185,6 +185,30 @@ void Consumer::drain()
         m_orch->doTask(*this);
 }
 
+string Consumer::dumpTuple(KeyOpFieldsValuesTuple &tuple)
+{
+    string s = getTableName() + getConsumerTable()->getTableNameSeparator() + kfvKey(tuple)
+               + "|" + kfvOp(tuple);
+    for (auto i = kfvFieldsValues(tuple).begin(); i != kfvFieldsValues(tuple).end(); i++)
+    {
+        s += "|" + fvField(*i) + ":" + fvValue(*i);
+    }
+
+    return s;
+}
+
+void Consumer::dumpToSyncTasks(vector<string> &ts)

I will investigate and hopefully submit a PR 
In reply to: 209787965 [](ancestors = 209787965)

**Comment by lguohan** [comment]
test vs fails, can you fix?

**Comment by lguohan** [comment]
retest this please
